### PR TITLE
Add AggressiveInline To ToDouble Methods

### PR DESCRIPTION
**Pull request by khalidabuhakmeh**
This helps with the performance of ToDouble as stated in issue #2106

### CODE REVIEW

**Comment by tannergooding** [comment]
cc. @ViktorHofer, there look to be a lot of unrelated failures that just popped up, is this a known issue?

**Comment by ViktorHofer** [comment]
Jep there's an outage. See mail offline.

**Comment by tannergooding** [comment]
Thanks for the contribution @khalidabuhakmeh, this LGTM but will require the outage to be resolved before tests can run.

I'll follow up and restart CI once the outage is complete 😄

**Comment by khalidabuhakmeh** [comment]
Aaaaaaah! I see that's a lot of red. Hope it's helpful :)

File: src/libraries/System.Private.CoreLib/src/System/BitConverter.cs [hunk 1]
```diff
@@ -608,6 +608,7 @@ public static float ToSingle(ReadOnlySpan<byte> value)
         /// </exception>
         /// <exception cref="ArgumentNullException"><paramref name="value"/> is null.</exception>
         /// <exception cref="ArgumentOutOfRangeException"><paramref name="startIndex"/> is less than zero or greater than the length of <paramref name="value"/> minus 1.</exception>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/BitConverter.cs:611]: Does this method really need it? This method look small enough to get inlined naturally.

File: src/libraries/System.Private.CoreLib/src/System/BitConverter.cs [hunk 2]
```diff
@@ -616,6 +617,7 @@ public static float ToSingle(ReadOnlySpan<byte> value)
         /// <param name="value">A read-only span containing the bytes to convert.</param>
         /// <returns>A double-precision floating-point value representing the converted bytes.</returns>
         /// <exception cref="ArgumentOutOfRangeException">The length of <paramref name="value"/> is less than 8.</exception>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/BitConverter.cs:620]: There are many other similar methods in this file: ToSingle, ToHalf, ... . Do all of them need this?

Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/BitConverter.cs:620]: Yes, we should also update the Single/Half versions, at a minimum.

File: src/libraries/System.Private.CoreLib/src/System/BitConverter.cs (hunk 1 above)
Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/BitConverter.cs:611]: I'm getting a JIT dump to check.

Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/BitConverter.cs:611]: Looks like this one is indeed small enough:
> INLINER: during 'fgInline' result 'success' reason 'below ALWAYS_INLINE size' for 'ConsoleApp50.Program:Main(System.String[])' calling 'System.BitConverter:ToDouble(System.Byte[],int):double'
> INLINER: during 'fgInline' result 'success' reason 'below ALWAYS_INLINE size'

The inner `ToInt64` fails to inline as it is not profitable.
```
Inline candidate looks like a wrapper method.  Multiplier increased to 1.
Inline candidate has an arg that feeds a constant test.  Multiplier increased to 2.
Inline candidate has arg that feeds range check.  Multiplier increased to 2.5.
Inline candidate callsite is boring.  Multiplier increased to 3.8.
calleeNativeSizeEstimate=801
callsiteNativeSizeEstimate=115
benefit multiplier=3.8
threshold=437
Native estimate for function size exceeds threshold for inlining 80.1 > 43.7 (multiplier = 3.8)
```

Which is a bit unfortunate as you might expect something like the following to be able to drop all the pre-checks:
```csharp
var bytes = new byte[1024];
new Random().NextBytes(bytes);
double d = BitConverter.Int64BitsToDouble(BitConverter.ToInt64(bytes, 0));
```

**Comment by tannergooding** [comment]
@khalidabuhakmeh, would you be able to respond to the remaining feedback?

There is just a method that shouldn't need the annotation and the Single/Half variants of the method that does get it should also get it.

**Comment by khalidabuhakmeh** [comment]
I removed the `[MethodImpl(MethodImplOptions.AggressiveInlining)]` from the `ToDouble` lambda based on the conversation that it has no effect on the performance.

**Review by tannergooding — APPROVED** [review]

**Comment by jkotas** [comment]
Could you please also add it to all other similar ToXXX methods that take Span? For example, `ToInt64` has the same problem.

**Pull request merged**